sklearn/feature_extraction/text.py

```diff
@@ -304,10 +305,35 @@ def _check_stop_words_consistency(self, stop_words, preprocess, tokenize):
             self._stop_words_id = id(self.stop_words)
             return 'error'
 
+    def _validate_analyzer(self):
+        # This is to check if the given custom analyzer expects file or a
+        # filename instead of data.
+        # Behavior changed in v0.21, function could be removed in v0.24
+        if not callable(self.analyzer):
+            return
+
+        import tempfile
+        with tempfile.NamedTemporaryFile() as f:
+            fname = f.name
+        # now we're sure fname doesn't exist
+
+        msg = ("Since v0.21, vectorizers pass the data to the custom analyzer "
+               "and not the file names or the file objects.")
+        try:
+            self.analyzer(fname)
+        except FileNotFoundError:
```

Comment:
I think we should just raise the warning in all cases of error.

Reply:
Same as above.